Add `profiler_path` option to `SCons` builds, with support for `tracy` and `perfetto`.

Ivorforce opened the pull request:
This is a minimal implementation to support the [tracy profiler](https://github.com/wolfpld/tracy/releases/tag/v0.11.1) and the [perfetto profiler](https://perfetto.dev).

<img width="1512" alt="SCR-20250331-qkvh" src="https://github.com/user-attachments/assets/faf8ebd2-3ad3-44e2-9d77-91373e6e6a59" />

<img width="1510" alt="SCR-20250331-trct" src="https://github.com/user-attachments/assets/3abf5055-79f5-4350-b461-8dd369d8192d" />


- Related to https://github.com/godotengine/godot-proposals/issues/5346

If this is accepted, future PRs could expand the granularity of frame tracking to include the various servers, steps etc.
Future PRs could also add support for other profilers (e.g. Perfetto).

## Why?

Tracy is a popular profiler for games and other "real time" software. "Profiling" means tracking performance over time, to try to figure out the cause of lag spikes, low frame rates, high RAM use, and others. The ultimate goal of course being to eliminate those problems.

It is already possible to profile Godot games. For one, Godot has a small [profiler built-in](https://docs.godotengine.org/en/stable/tutorials/scripting/debug/the_profiler.html). However, it's lacking features for serious profiling tasks, especially to profile c++ code.

There are [guides](https://docs.godotengine.org/en/stable/contributing/development/debugging/using_cpp_profilers.html) to use external C++ profilers on the docs as well. These are capable of a lot, but there are features offered by tracy that are not supported by other profilers - for example per-frame profiling, which requires injection of a high precision callback to be worthwhile.

Here are the supported features per platform, stolen from the current [tracy manual](https://github.com/wolfpld/tracy/releases/download/v0.11.1/tracy.pdf):
<img width="866" alt="SCR-20250402-bhxh" src="https://github.com/user-attachments/assets/abf06f5c-988c-4878-b25f-0732db105813" />

## Alternatives

I am not familiar with any profilers except `Instruments` (including `tracy`). This PR is a first step to get familiar with them.
The idea is to establish a framework that allows injection of any useful intrusive profilers. For example, `Perfetto` may be integratable too, if desired by someone.

It is currently also possible to integrate tracy support with the [Godot Tracy Module](https://github.com/AndreaCatania/godot_tracy) (cc @AndreaCatania). This is a great effort, and I've referenced this for my implementation.
The reason I'm still proposing to integrate this into the engine itself is that intrusive changes to the codebase are required to support frame tracking, which is arguably the most useful feature of tracy. It would be nice for the engine to 'just' support this.

## Implementation notes

This is the first module I am working with / proposing. Chances are the approach is suboptimal. Please suggest improvements! :)

In particular, my idea was that you could include `profiling.h` and use generic macros to mark things to profile, regardless of the actual profiler used (in assumption that others work similar to `tracy` in this regard).

## License
Tracy uses the [`BSD` license](https://github.com/wolfpld/tracy/blob/master/LICENSE), which is compatible with ours (thanks @AThousandShips and @Calinou!)

AThousandShips commented:
You need to add this to `COPYRIGHT.md` and to `thirdparty/README.md`

AThousandShips commented:
This would also be best tracked in a dedicated proposal, especially to compare alternatives and options

clayjohn commented:
This is awesome! I actually looked at integrating Tracy earlier this month, but then decided not to continue because of all the required build system stuff to make it work. 

Ideally this module would function in a profiler agnostic way. Perfetto is really nice and is also open source (its only 2 files IIRC so may be a better fit for a default profiler). A lot of people also like Pix which is Microsoft's tool for PC/Xbox. 

I have discussed with a few people over the last year or so the need to be able to drop in a profiler (or have an integrated profiler with a way to easily replace markers with something else). That way we can have timestamps/markers that stay in the engine and we can track performance over time, and users of the profilers don't need to constantly re-add markers with every engine update.

Calinou commented:
> Tracy uses the [BSD license](https://github.com/wolfpld/tracy/blob/master/LICENSE?rgh-link-date=2025-03-31T17%3A00%3A13Z).
I think it's compatible, but I'm no expert. So please weigh in if you know for sure!

Yes, it's compatible :slightly_smiling_face:

Ivorforce commented:
> BEGIN_ZONE / END_ZONE

I was looking for something like this, I haven't found it in tracy so far. Currently, I'm just adding empty scopes, which will not change function behavior, but it does incur some (perhaps unnecessary) whitespace changes. At the same time, usually you'll just want to annotate a full function, so it's probably fine?

It may also be possible to create manual start and end macros _for_ tracy by wrapping its `RTTI` behavior and manually destructing the object. Perhaps even `REPLACE_ZONE` for as a convenient end + begin.

Ivorforce commented:
> Is it technically possible to allow declaring Tracy sections from the scripting API, so you can profile GDScript/C# code (or at least the C++ side of the methods being called within that block)?

I think so! 
Tracy supports runtime-defined zones through "transient" zones, which are but a tad slower / more memory intensive than hard zones. It should also be possible to enable automated stack sampling only for specific transient zones (for performance reasons).
The godot tracy module project has an implementation for this: https://github.com/AndreaCatania/godot_tracy/blob/main/tracy_profiler.cpp

Tracy has really low overhead, so it may even be an option to even profile all GDScript functions 'by default' (when using a tracy profile build). An interface could still be used to profile specific subsections of a function.

Ivorforce committed:
Add `profiler` option to `SCons` builds.
    Add `tracy` option to `profiler`. If set, a tracy profiling client will be injected into the Godot binary.

# Conflicts:
#	platform/linuxbsd/godot_linuxbsd.cpp

Ivorforce committed:
Add some important profiling hooks.

Ivorforce commented:
> I do not see a way to annotate code that would not be scoped.

I managed to do it! It's not possible with tracy plain, but with some object lifetime hacking, it can be achieved.
It's still all scoped (which is probably a good idea for sanity reasons), but it's possible now to define "grouped" sections to divide up the larger functions like `Main::iteration`. This allowed me to revert the many whitespace changes, and makes profiling sections _very_ unintrusive.

Ivorforce commented:
I have removed `tracy` code from the PR, as discussed in the core meeting. tracy now has to be placed manually in `modules/profiler/tracy`. 
To give some background, this allows users to select the tracy version manually (provided it didn't change interfaces recently), and relieves us from needing to host / copy the source code.

Calinou commented:
> tracy now has to be placed manually in modules/profiler/tracy.

It's a good idea to include a downloader script similar to https://github.com/godotengine/godot/blob/master/misc/scripts/install_d3d12_sdk_windows.py (e.g. `setup_tracy.py`), which would download the latest release ZIP and extract it to the right folder. This saves having to do it manually and is handy for CI, as we'll want to build with Tracy enabled to avoid regressions.

Ivorforce commented:
> > tracy now has to be placed manually in modules/profiler/tracy.
> 
> It's a good idea to include a downloader script similar to https://github.com/godotengine/godot/blob/master/misc/scripts/install_d3d12_sdk_windows.py (e.g. `misc/scripts/install_tracy.py`), which would download the latest release ZIP and extract it to the right folder. This saves having to do it manually and is handy for CI, as we'll most likely want to build with Tracy enabled to avoid undetected regressions.
> 
> The same goes for Perfetto if it needs a copy of its source code.

This idea was brought up in the meeting as well.
The idea has some merit, although both perfetto and tracy need external setup anyway so I'm not sure how useful this automation would be.
In any case, I think it might be better to discuss this for a follow-up PR.

kisg commented:
As the original author of the related proposal (and a 3.x perfetto-only version) thank you for your work on this! I hope it gets merged soon for everyone's benefit.

Calinou commented on modules/profiler/SCsub:
```diff
@@ -0,0 +1,100 @@
+opts.Add(
+    EnumVariable(
+        "profiler",
+        "Profiler framework to use compile into the binary.",
```
The SCons option descriptions we have elsewhere don't end with a period:

```suggestion
        "Profiler framework to use compile into the binary",
```

Calinou commented on modules/profiler/SCsub:
```diff
@@ -0,0 +1,100 @@
+opts.Add(
+    BoolVariable(
+        "profiler_sample_callstack",
+        "Profile random samples application-wide using a callstack based sampler.",
```
```suggestion
        "Profile random samples application-wide using a callstack based sampler",
```

Calinou commented:
While compiling with a symlink, I got this warning:

```
WARNING: Failed to redirect "/home/hugo/Documents/Git/wolfpld/tracy/public/TracyClient.linuxbsd.editor.x86_64.o"
```

I double-checked my symlink resolves correctly. The file exists and is 2.6 MB large.

I've never used Tracy before, so I don't know how you start its GUI on Linux. I've tried self-compiling it via CMake and the default options, but I don't see any program binary being compiled, only a static library (or a dynamic library when Godot compiles Tracy from SCons).

PS: Should we enable Tracy support in official binaries? As I understand it, debug symbols are not a strict requirement for Tracy to work (even if it's not as precise without debug symbols). This would allow users to troubleshoot some performance issues without needing to self-compile Godot.

Ivorforce commented:
> While compiling with a symlink, I got this warning:
> 
> ```
> WARNING: Failed to redirect "/home/hugo/Documents/Git/wolfpld/tracy/public/TracyClient.linuxbsd.editor.x86_64.o"
> ```

I suppose this is somewhat of a regression of #101641 (cc @Repiteo). I think it should be fixable if we don't resolve symlinks in emitters.

> I've never used Tracy before, so I don't know how you start its GUI on Linux. I've tried self-compiling it via CMake and the default options, but I don't see any program binary being compiled, only a static library (or a dynamic library when Godot compiles Tracy from SCons). 

Yea, I had trouble compiling it from scratch too. In the end I just used `brew install tracy`. Perhaps there's pre-built binaries for Linux somewhere too?

> The manual isn't really explicit on how you're supposed to start a profiling-enabled session, collecting data then viewing it. I'm much more used to the HotSpot workflow in comparison.

Once you have the GUI it's pretty easy; you just compiler with `scons profiler=tracy` and run Godot. On the tracy side, you just run `tracy`, and in the GUI, click the `connect` button.

> PS: Should we enable Tracy support in official binaries? As I understand it, debug symbols are not a strict requirement for Tracy to work (even if it's not as precise without debug symbols). This would allow users to troubleshoot some performance issues without needing to self-compile Godot.

Technically tracy is so low overhead that it would probably be OK to have it in official builds by default. On the other hand, there's cons like needing to include tracy source code after all, matching the tracy version, and some people would probably argue `perfetto` should be the default.
At the very least it's probably something we'd want to discuss after having the framework in place.

Repiteo commented:
Shoot, I didn't even *consider* symlink shenanigans in the emitter. I've been meaning to revisit the emitter anyway to remove hardcoded paths, but now I have a real motivator!

Calinou commented:
@Ivorforce Thanks! That works as long as I also add `-DLEGACY=1` to the first `cmake` command, so that it can start on X11 (otherwise, only Wayland support is enabled).

Calinou commented:
It works now, but I get a link error if I build without `profiler=tracy` after building with it once:

```
scons: Reading SConscript files ...
SCU: Generating build files... (max includes per SCU: 8)
Using linker program: mold
WARNING: System-provided icu4c or harfbuzz cause known issues for GDExtension (see GH-91401 and GH-100301).
Building for platform "linuxbsd", architecture "x86_64", target "editor".
Checking for C header file mntent.h... (cached) yes
scons: done reading SConscript files.
scons: Building targets ...
Linking Program bin/godot.linuxbsd.editor.x86_64 ...
mold: error: undefined symbol: tracy::GetProfilerv
>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(OS_LinuxBSD::runv)>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(OS_LinuxBSD::runv)>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(OS_LinuxBSD::runv)>>> referenced 63 more times

mold: error: undefined symbol: tracy::rpmallocm
>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(tracy::tracy_malloc(unsigned long))>>> referenced by scu_servers_rendering_1.gen.cpp
>>>               bin/obj/servers/libservers.linuxbsd.editor.x86_64.a(scu_servers_rendering_1.gen.linuxbsd.editor.x86_64.o):(tracy::tracy_malloc(unsigned long))>>> referenced by main.cpp
>>>               bin/obj/main/libmain.linuxbsd.editor.x86_64.a(main.linuxbsd.editor.x86_64.o):(tracy::tracy_malloc(unsigned long))>>> referenced 2 more times

mold: error: undefined symbol: tracy::InitRpmallocv
>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(tracy::tracy_malloc(unsigned long))>>> referenced by scu_servers_rendering_1.gen.cpp
>>>               bin/obj/servers/libservers.linuxbsd.editor.x86_64.a(scu_servers_rendering_1.gen.linuxbsd.editor.x86_64.o):(tracy::tracy_malloc(unsigned long))>>> referenced by main.cpp
>>>               bin/obj/main/libmain.linuxbsd.editor.x86_64.a(main.linuxbsd.editor.x86_64.o):(tracy::tracy_malloc(unsigned long))>>> referenced 2 more times

mold: error: undefined symbol: tracy::GetTokenv
>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(OS_LinuxBSD::runv)>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(OS_LinuxBSD::runv)>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(OS_LinuxBSD::runv)>>> referenced 177 more times

mold: error: undefined symbol: tracy::rpfreePv
>>> referenced by os_linuxbsd.cpp
>>>               bin/obj/platform/linuxbsd/os_linuxbsd.linuxbsd.editor.x86_64.o:(tracy::tracy_free_fast(void*))
collect2: error: ld returned 1 exit status
scons: *** [bin/godot.linuxbsd.editor.x86_64] Error 1
scons: building terminated because of errors.
```

![Screenshot_20250412_164154 png webp](https://github.com/user-attachments/assets/de999a1d-8694-4791-90a7-ef60a1de3644)

Ivorforce commented:
> It works now, but I get a link error if I build without profiler=tracy after building with it once:

I noticed that too. I think it's because of caching / dependencies of generated files. Gotta figure that out before merging.

kisg commented on servers/rendering/renderer_viewport.cpp:
```diff
@@ -753,6 +757,8 @@ void RendererViewport::draw_viewports(bool p_swap_buffers) {
+
 	//determine what is visible
 	draw_viewports_pass++;
 
```
Again, can't put this one to the right line, but we should have a separate slice for each rendered viewport:

```cpp
	for (int i = 0; i < sorted_active_viewports.size(); i++) {
		GodotProfileZone("render viewport");
		Viewport *vp = sorted_active_viewports[i];
```

kisg reviewed (changes requested):
Overall, I think this PR is a great first step in properly instrumenting Godot to easily gather detailed profiling data. 

With the changes I propose we were already able to find a possible performance bottleneck in an Android project we are working on.

Ivorforce commented on core/profiling.h:
```diff
@@ -0,0 +1,141 @@
+	PerfettoGroupedEventEnder __godot_perfetto_zone_##m_group_name("general")
+#define GodotProfileZoneGroupedEndEarly(m_group_name, m_zone_name) __godot_perfetto_zone_##m_group_name.~PerfettoGroupedEventEnder()
+#define GodotProfileZoneGrouped(m_group_name, m_zone_name) \
+	TRACE_EVENT_BEGIN("general", m_zone_name);             \
```
This is great, thanks!

I wanted to avoid re-inventing the RAII pattern for this, since they obviously are doing a lot of extra work in their own RAII pattern. But if this works better, I'm happy to change it!

Ivorforce commented on core/profiling.h:
```diff
@@ -0,0 +1,141 @@
+	TRACE_EVENT_BEGIN("general", m_zone_name);             \
+	__godot_perfetto_zone_##m_group_name.change("general")
+
+void godot_init_profiler();
```
Generally I agree. However, I think it's out of scope to do this now, so I'd rather keep it for a follow-up PR.

Ivorforce commented on main/main.cpp:
```diff
@@ -4631,6 +4638,7 @@ bool Main::iteration() {
 		PhysicsServer2D::get_singleton()->flush_queries();
 #endif // PHYSICS_2D_DISABLED
 
```
Good ideas! 
I was hoping to expand this eventually anyway, either now or in follow-up PRs. If you've spent some time thinking about this I'm happy to expand now.

Ivorforce commented:
Many thanks to @kisg!
I've pushed the suggested changes. I've also fixed the `profiler` module not realizing when the profiler changes, to re-generate `profiling.gen.h`. This should fix the problem where `profiling.gen.h` has to be manually deleted to switch profiler configurations.

ze2j commented:
Supporting Tracy would be a great addition to the engine. I integrated Tracy into my fork about a year ago, and it’s been incredibly helpful for optimizing my game — and even [the engine itself](https://github.com/godotengine/godot/pull/98760).

However, since Tracy is statically linked to the executable, it can’t be used to profile `GDExtension`s written in C++, even though they could benefit from it as well. In my custom fork, I built Tracy as a shared library and linked it both to the game executable and to the GDExtension shared library (as explained by the Tracy's documentation). It works, but the setup is a bit more involved (LD_LIBRARY_PATH...) compared to this PR.

@Ivoforce Do you plan to support `GDExtension` as well? I’m concerned that a module-based implementation might make it difficult — or even impossible. I know modules can be built as shared libraries, but I’m not sure if linking a module to a `GDExtension` would be possible or even make sense.

Ivorforce commented:
I've moved the profiler to an `SCSub` inside `main/profiling`, as (loosely) suggested above. I think this is more appropriate than the previous solution (as a module), since the profiler integrates tightly into Godot, and isn't very isolatable - especially if we try to consolidate this new profiling code with what's floating around the codebase already.

m4gr3d commented on main/profiling/SCsub:
```diff
@@ -0,0 +1,98 @@
+        Exit(255)
+    if not tracy_dir_path.is_dir():
+        print(
+            "'tracy/public/' folder is present, but 'tracy/public/' is missing. Please make sure you have copied or symlinked 'tracy/' as 'modules/profiler/tracy/'."
```
```suggestion
            "'tracy/' folder is present, but 'tracy/public/' is missing. Please make sure you have copied or symlinked 'tracy/' as 'modules/profiler/tracy/'."
```

m4gr3d commented on platform/android/java_godot_lib_jni.cpp:
```diff
@@ -165,6 +166,8 @@ JNIEXPORT void JNICALL Java_org_godotengine_godot_GodotLib_ondestroy(JNIEnv *env
 }
 
 JNIEXPORT jboolean JNICALL Java_org_godotengine_godot_GodotLib_setup(JNIEnv *env, jclass clazz, jobjectArray p_cmdline, jobject p_godot_tts) {
+	godot_init_profiler();
```
This should be invoked in `initialize` to match the pattern used for the other platforms.

Repiteo commented:
> Could you explain why you moved the options to `SConstruct`, but kept the logic in `profiling/SCsub`?

Because adding additional construction arguments is currently relegated to modules and platforms, which support means of checking if the feature is supported to begin with. Your integration updates the options in a later step, which is something we currently have to *explicitly* avoid in our current SCons setup. It'd be a better location if our buildsystem was setup to support it, but that's outside the scope of this PR.

> And why the `profiler_gen_builder` is moved to `profiling_builders.py `? 

Because one of the first changes to a buildsystem overhaul will involve making generator code decoupled from build code. So while it's not strictly necessary, I'm making an effort to have what generator code we currently have exist outside of SCons files where possible, so the changes can be mostly atomic when that switch happens.

> I've also noticed that with my current setup, the tracy paths aren't added to all compile commands, just some of them. It seems to be covering all that's needed at the moment, but it's still somewhat worrying, because I think it means there might be files that cannot use the profilers. I'm not sure what causes this, but possibly it's because it's called too late in `SConstruct`, when some environments have already been cloned off. I'm open to suggestions for how best to fix this (move `SCSub` out of main / call it earlier in the main `SConstruct`?).

`/main/SCsub` is the penultimate call in the `SConstruct` logic chain, with only `/platforms/<target>/SCsub` called after[^1]. This is also **meant** to represent the order in which encapsulation is permitted (i.e. nothing from `scene` should be used by `core`), but isn't strictly enforced (and likely the biggest cause of buildtime regressions compared to 3.x).

With that in mind, this logic should be migrated to `/core/profiling`. This would ensure it gets parsed at the very start & respects the encapsulation logic. Nothing in the profiler scripts use non-thirdparty headers outside of `/core`, so it would be an effortless transition.

Now, it's *possible* that won't actually be sufficent. I don't know for certain if the preceeding `Export("env")` step implies that the environment needs to be re-imported & re-exported after `SConscript("core/SCsub")` is processed. If it comes to that, this logic could be better suited for the `detect.py` scripts of platforms, which occur before the subdirectory steps. That technically could be done with modules too, but that might be relegated to `env_modules`, I'll have to double-check.

If any of this sounds hacky or arbitrary… Yeah.

[^1]: https://github.com/godotengine/godot/blob/d705613db3752115f5b1244494bde6471fe6700d/SConstruct#L1175-L1188

Ivorforce commented:
Thanks for the explanations @Repiteo.

I've moved `profiling` to `core` (from `main`) and moved the options to `SConstruct`. 
I've also made the `profiling_builders.py` change, although I'm not quite show how this will be callable without SCons integration? The function calls are pretty tightly coupled. But anyway, this is not really related to this PR.

mhilbrunner reviewed (approved):
Huge QOL addition for everyone to not have to add this manually (again and again). I think this ended up at a good approach code wise, it looks like a fairly small and easy to use addition. Maintainable, which is always nice :)

I tested this on a Linux and a Windows box and both seem to work as advertised, for example just adding `profiler_path="path/to/profiler-src/"`. Great work!

Ivorforce commented on servers/rendering/rendering_device.cpp:
```diff
@@ -6256,12 +6257,16 @@ String RenderingDevice::get_device_pipeline_cache_uuid() const {
 void RenderingDevice::swap_buffers(bool p_present) {
 	ERR_RENDER_THREAD_GUARD();
 
+	GodotProfileZoneGroupedFirst(_profile_zone, "_end_frame");
```
Yea, generally I've been preferring the `Class::function` when it's meaningful, but in some cases it wasn't, especially when the zone was just part of some arbitrary group. It would be nice to find a pattern to use, but currently it's just the "Ivorius has been doing his best" pattern 😅

Ivorforce commented on servers/rendering/renderer_viewport.cpp:
```diff
@@ -821,6 +825,9 @@ void RendererViewport::draw_viewports(bool p_swap_buffers) {
 	int draw_calls_used = 0;
 
 	for (int i = 0; i < sorted_active_viewports.size(); i++) {
+		// TODO Somehow print the index
```
Yea, I'm hoping we can expand the API to add support for indices in both tracy and perfetto in a future PR>

batesj commented on servers/rendering/renderer_viewport.cpp:
```diff
@@ -750,6 +752,7 @@ void RendererViewport::draw_viewports(bool p_swap_buffers) {
 	}
 
 	if (sorted_active_viewports_dirty) {
+		GodotProfileZoneGrouped(_profile_zone, "_sort_active_viewports");
```
I was thinking it is unexpected that this _sort_active_viewports scope continues beyond the end of this if statement (similar to my comment above). Since it's called "_sort_active_viewports" I guess it should only trace this call to _sort_active_viewports(), and GodotProfileZone does exactly that (at least, that's how things would work with the perfetto UI -- this GodotProfileZone would be stacked underneath the parent  "prepare viewports" event).

Ivorforce replied:
Stacking `_sort_active_viewports` below `prepare viewports` would be possible, if this statement used `GodotProfileZone` instead, as you suggested.
However, in this case this is not wanted, because `prepare viewports` doesn't (in my mind) include sorting the viewports. It's a different step.
Ending the `_sort_active_viewports` might be nice, but I'd prefer to investigate that in a follow-up PR. It would require adding a _new_ object on every `GodotProfileZoneGrouped` call, and ending the grouped scope when it is destructed, which might result in a lot of added complexity.

Ivorforce commented on main/main.cpp:
```diff
@@ -4827,6 +4853,7 @@ bool Main::iteration() {
 	iterating--;
 
 	if (movie_writer) {
+		GodotProfileZoneGrouped(_profile_zone, "movie_writer->add_frame");
```
Yea, I see the problem. I think the correct solution would be to re-write the perfetto macros to work like the tracy ones.
I can probably do this with a Godot helper, but I'd need to dive into the APIs more. And as mentioned earlier, I currently can't run perfetto, because it's refusing to cooperate with me.

As much as I'd love to solve this, I think I'd prefer to merge the PR, so we can get to work correcting the implementations later. The important part of this PR is the API design; the implementation can always be improved later.

Repiteo reviewed (approved):
I'm satisfied with the implementation as-is, and would love to lay the groundwork for these tools to work natively within the engine. The SCons accommodations are great, thanks for taking care of that!

enetheru commented:
> I've moved the profiler to an `SCSub` inside `main/profiling`, as (loosely) suggested above. I think this is more appropriate than the previous solution (as a module), since the profiler integrates tightly into Godot, and isn't very isolatable - especially if we try to consolidate this new profiling code with what's floating around the codebase already.

I was working on adapting the things I have working to your PR.
It appears to me that the profiler_path option is added after the help text is generated, making it not show up in the final result. 
given the steam api and tests in the main/SCsub are in the main SConstruct wouldnt the profiler_path option also belong there?

enetheru commented:
> I think that should have fixed the issues you were experiencing. Did you test the latest version?

oh it looks like my branch got out of sync with yours. i'll fix it now, cheers.
